### PR TITLE
build-machinery: Add version support

### PR DESCRIPTION
**Pull request by tnozicka**
/cc @deads2k @sttts

### CODE REVIEW

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
 all:
 	$(MAKE) -C . build
 	[[ -f ./openshift ]]
 	[[ -f ./oc ]]
 
+	# test version is set correctly when linking
+	#   majorFromGit, minorFromGit are deprecated upstream and set to empty value
+	#   we avoid comparing time to avoid flakes
+	set -euo pipefail; diff <( ./oc | sed '$$d' ) <( \
+			echo ""; \
```
Comment (sttts) [inline alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: odd deep indention

**Comment by tnozicka** [comment]
got the bash issue sorted out the unit test is OOM (fix https://github.com/openshift/release/pull/4162)

**Comment by tnozicka** [comment]
green now

File: alpha-build-machinery/Makefile [hunk 2]
```diff
@@ -11,7 +11,9 @@ examples :=$(wildcard ./make/examples/*/Makefile.test)
 # We need to change dir to the final makefile directory or relative paths won't match
 define update-makefile-log
 mkdir -p "$(3)"
-$(MAKE) -C "$(dir $(1))" -f "$(notdir $(1))" --no-print-directory --warn-undefined-variables $(2) 2>&1 | tee "$(3)"/"$(notdir $(1))"$(subst ..,.,.$(2).log)
+$(MAKE) -j 1 -C "$(dir $(1))" -f "$(notdir $(1))" --no-print-directory --warn-undefined-variables $(2) 2>&1 | \
```
Comment (sttts) [inline alpha-build-machinery/Makefile]: what does this do? needs a comment

**Comment by sttts** [comment]
Do we have a README describing these features?

File: alpha-build-machinery/Makefile (hunk 2 above)
Comment (tnozicka) [reply alpha-build-machinery/Makefile]: `-j 1` means serial (jobs) - so it is deterministic in tests and the diffs match

**Comment by tnozicka** [comment]
we can do README when we mature from alpha

**Comment by tnozicka** [comment]
(reminder we need to bump this into `origin` before freeing staging repos)

File: alpha-build-machinery/Makefile (hunk 2 above)
Comment (sttts) [reply alpha-build-machinery/Makefile]: I meant the whole 3 line monster, especialy the sed

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test [hunk 3]
```diff
@@ -1,8 +1,21 @@
...
 	[[ -f ./openshift ]]
 	[[ -f ./oc ]]
 
+	# test version is set correctly when linking
+	#   majorFromGit, minorFromGit are deprecated upstream and set to empty value
+	#   we avoid comparing time to avoid flakes
+	diff <( ./oc | sed '$$d' ) <( \
+		echo ""; \
+		echo ""; \
+		echo $$( git rev-parse --short "HEAD^{commit}" 2>/dev/null ); \
```
Comment (sttts) [inline alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: why echo and not directly the git command?

Comment (sttts) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: Same everywhere else.

File: alpha-build-machinery/Makefile (hunk 2 above)
Comment (tnozicka) [reply alpha-build-machinery/Makefile]: the sed replaces stuff like the build date (which differs every invocation)  with a string "<redacted_for_diff>" or the checkout templates would never match

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test (hunk 3 above)
Comment (tnozicka) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: say `rev-parse` returns `4d798104` then you'd get `/bin/bash: 4d798104: command not found`

Comment (sttts) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: no. Just remove ech and the `$$( ... )`

Comment (sttts) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]:
```
diff <( ./oc | sed '$$d' ) <( \
		echo ""; \
		echo ""; \
		git rev-parse --short "HEAD^{commit}" 2>/dev/null; \
...
)
```

Comment (tnozicka) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test]: sorry, I forgot to remove the subshell when trying your suggestion

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log [hunk 4]
```diff
@@ -1,8 +1,21 @@
...
 [[ -f ./oc ]]
+# test version is set correctly when linking
+#   majorFromGit, minorFromGit are deprecated upstream and set to empty value
+#   we avoid comparing time to avoid flakes
+diff <( ./oc | sed '$d' ) <( \
+	echo ""; \
+	echo ""; \
+	git rev-parse --short "HEAD^{commit}" 2>/dev/null; \
+	git describe --long --tags --abbrev=7 --match 'v[0-9]*' || echo 'v0.0.0-unknown'; \
+	(git diff --quiet && echo 'clean') || echo 'dirty'; \
```
Comment (sttts) [inline alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log]: `x && y || z` will do what you want.

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log [hunk 5]
```diff
@@ -1,8 +1,21 @@
...
+)
+fatal: No names found, cannot describe anything.
```
Comment (sttts) [inline alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log]: echo?

Comment (sttts) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log]: oh, wrong patch. This is the log. Nvmd.

File: alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log (hunk 4 above)
Comment (tnozicka) [reply alpha-build-machinery/make/examples/multiple-binaries/Makefile.test.log]: parenthesis looked better to me, but why not

**Pull request merged**